src/database.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provide the DatabaseObserver class to handle database relation and state."""
+
+import typing
+
+from charms.data_platform_libs.v0.data_interfaces import DatabaseRequires
+from ops.charm import CharmBase
+from ops.framework import Object
+
+from constants import DATABASE_NAME
+
+class DatabaseService(Object):
+    """The Database relation observer."""
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Initialize the oserver and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        super().__init__(charm, "database-observer")
+        self._charm = charm
+        self.relation_name = relation_name
+        self.database = DatabaseRequires(
+            self._charm,
+            relation_name=self.relation_name,
+            database_name=DATABASE_NAME,
+        )
+
+    def get_relation_data(self) -> typing.Dict[str, str]:
+        """Get database data from relation.
+
+        Returns:
+            Dict: Information needed for setting environment variables.
+            Returns default if the relation data is not correctly initialized.
+        """
+        default = {
+            "POSTGRES_USER": "",
+            "POSTGRES_PASSWORD": "",
+            "POSTGRES_HOST": "",
+            "POSTGRES_PORT": "",
+            "POSTGRES_DB": "",
+        }
+
+        if self.model.get_relation(self.relation_name) is None:
+            return default
+
+        relation_id = self.database.relations[0].id
+        relation_data = self.database.fetch_relation_data()[relation_id]
+
+        endpoints = relation_data.get("endpoints", "").split(",")
+        if len(endpoints) < 1:
+            return default
+
+        primary_endpoint = endpoints[0].split(":")
+        if len(primary_endpoint) < 2:
+            return default
```

Comment:
This silently returns while there are few chances it will change, I think it's better to return a validation error which will lead to blocked state